src/protocol/sort/shuffle.rs

```diff
@@ -1,97 +1,226 @@
+    async fn reshare_all_shares<M, G>(
+        &self,
+        ctx: &ProtocolContext<'_, G, S>,
+        which_step: ShuffleStep,
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, S>,
+    {
+        let step = (self.step_fn)(which_step);
+        let to_helper = Self::shuffle_for_helper(which_step);
+        let reshares = self
+            .input
+            .iter()
+            .enumerate()
+            .map(|(index, input)| async move {
+                Reshare::new(*input)
+                    .execute(ctx, RecordId::from(index as u32), step, to_helper)
+                    .await
+            });
+        try_join_all(reshares).await
+    }
+
+    /// `single_shuffle` is called for the helpers
+    /// i)   2 helpers generate random sequence using shared prss random number
+    /// ii)  2 helpers apply the permutation to their shares
+    /// iii) reshare to `to_helper`
+    async fn single_shuffle<M: Mesh, G: Gateway<M, S>>(
+        &mut self,
+        ctx: &ProtocolContext<'_, G, S>,
+        which_step: ShuffleStep,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let to_helper = Self::shuffle_for_helper(which_step);
+        let step = (self.step_fn)(which_step);
+        let prss = &ctx.participant[step];
+        let channel = ctx.gateway.get_channel(step);
+
+        if to_helper != channel.identity() {
+            let direction = if to_helper.peer(Direction::Left) == channel.identity() {
+                Direction::Left
+            } else {
+                Direction::Right
+            };
+            let mut permute = Self::generate_random_permutation(self.input.len(), direction, prss);
+            apply_inv(&mut permute, &mut self.input);
+        }
+        self.reshare_all_shares(ctx, which_step).await
+    }
+    #[embed_doc_image("shuffle", "images/sort/shuffle.png")]
+    /// Shuffle calls `single_shuffle` three times with 2 helpers shuffling the shares each time.
+    /// Order of calling `single_shuffle` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
+    /// Each single shuffle requires communication between helpers to perform reshare.
+    /// Infrastructure has a pre-requisite to distinguish each communication step uniquely.
+    /// For this, we have three shuffle steps one per `single_shuffle` i.e. Step1, Step2 and Step3.
+    /// The Shuffle object receives a step function and appends a `ShuffleStep` to form a concrete step
+    /// ![Shuffle steps][shuffle]
+    #[allow(dead_code)]
+    pub async fn execute<M: Mesh, G: Gateway<M, S>>(
+        &mut self,
+        ctx: &ProtocolContext<'_, G, S>,
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        F: Field,
+    {
+        self.input = self.single_shuffle(ctx, Step1).await.unwrap();
+        self.input = self.single_shuffle(ctx, Step2).await.unwrap();
+        self.input = self.single_shuffle(ctx, Step3).await.unwrap();
+
+        Ok(self.input.clone())
```

Comment:
@akoshelev is there a way to get rid of this clone? I tried making input mutable but since it has a lifetime attached to it, when we assign a new value to the input, the code doesn't compile

Reply:
Thanks for the offline suggestions! Fixed this